src/protocol/basics/mul/sparse.rs

```diff
@@ -0,0 +1,428 @@
+            if flags[role.peer(Right) as usize] {
+                assert_eq!(
+                    F::ZERO,
+                    v.right(),
+                    "expected a zero on the right for input {which}"
+                );
+            }
+        }
+    }
+
+    #[must_use]
+    pub fn all() -> &'static [ZeroPositions] {
+        const ALL: &[ZeroPositions] = &[
+            ZeroPositions::Pzvv,
+            ZeroPositions::Pvzv,
+            ZeroPositions::Pvvz,
+            ZeroPositions::Pvzz,
+            ZeroPositions::Pzvz,
+            ZeroPositions::Pzzv,
+            ZeroPositions::Pvvv,
+        ];
+        ALL
+    }
+}
+
+// Code for testing and debugging only.
+impl From<ZeroPositions> for [bool; 3] {
+    fn from(zp: ZeroPositions) -> Self {
+        match zp {
+            ZeroPositions::Pvvv => [false, false, false],
+            ZeroPositions::Pzzv => [true, true, false],
+            ZeroPositions::Pzvz => [true, false, true],
+            ZeroPositions::Pvzz => [false, true, true],
+            ZeroPositions::Pvvz => [false, false, true],
+            ZeroPositions::Pvzv => [false, true, false],
+            ZeroPositions::Pzvv => [true, false, false],
+        }
+    }
+}
+
+/// This struct includes public constants for different arrangements of known zeros in shares.
+/// The value of this can be used to determine which of the three helpers sends during a multiplication.
+pub(super) trait MultiplyWork {
+    /// Determine the work that is required for the identified role.
+    /// Return value is who is sending relative to the given role [self, left, right].
+    fn work_for(self, role: Role) -> [bool; 3];
+    /// Determines where there are known zeros in the output of a multiplication.
+    fn output(self) -> ZeroPositions;
+}
+
+impl MultiplyWork for MultiplyZeroPositions {
+    fn work_for(self, role: Role) -> [bool; 3] {
+        let work = ZeroPositions::work(self);
+        let i = role as usize;
+        let need_to_recv = work[i % 3];
+        let need_to_send = work[(i + 1) % 3];
+        let need_random_right = work[(i + 2) % 3];
+        [need_to_recv, need_to_send, need_random_right]
+    }
+
+    fn output(self) -> ZeroPositions {
+        ZeroPositions::mul_output(self)
+    }
+}
+
+#[cfg(test)]
+pub(in crate::protocol) mod test {
+    use crate::{
+        ff::{Field, Fp31, Fp32BitPrime},
+        helpers::{
+            Direction::{Left, Right},
+            Role,
+        },
+        protocol::{
+            basics::{mul::sparse::MultiplyWork, MultiplyZeroPositions, SecureMul, ZeroPositions},
+            malicious::MaliciousValidator,
+            BitOpStep, QueryId, RecordId,
+        },
+        rand::{thread_rng, Rng},
+        secret_sharing::Replicated,
+        test_fixture::{IntoShares, Reconstruct, Runner, TestWorld},
+    };
+    use rand::distributions::{Distribution, Standard};
+    use std::{borrow::Borrow, iter::zip};
+
+    #[derive(Clone, Copy)]
+    pub struct SparseField<F> {
```

Comment:
Sorry to quibble about the name, but it's not really a sparse _Field_, so much a sparse _Sharing_, right?

Reply:
FieldWithSparseShares ?  It is still a field, but when it turns into shares it produces fewer values...  I really don't know what the right framing is here.